### DISCUSSION TITLE
Turn on NULL cipher

### DISCUSSION THREAD

**Discussion by ClemFandango2** [openssl/openssl#22144] [Q&A] [answered]
For learning purposes, I want to turn on NULL cipher.
So that no encryption would go on.
And so I could record all the traffic using Wireshark.

I've found that I can start a https webserver, toyish:

```
openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -days 365 -nodes
openssl s_server -key key.pem -cert cert.pem -accept 44330 -www -cipher NULL-SHA
```

I'm running lynx:
```
% lynx https://localhost:44330
...
Ciphers supported in s_server binary
TLSv1.3    :TLS_AES_256_GCM_SHA384    TLSv1.3    :TLS_CHACHA20_POLY1305_SHA256
TLSv1.3    :TLS_AES_128_GCM_SHA256    SSLv3      :NULL-SHA
...
```

Seems that server supports NULL-SHA?
But how to connect to it and record all the (unencrypted) traffic in Wireshark?
That doesn't work:
```
% openssl s_client -cipher "NULL-SHA" -no_tls1_3 -connect localhost:44330
...
CONNECTED(00000003)
4087A3F3777F0000:error:0A0000B5:SSL routines:ssl_cipher_list_to_bytes:no ciphers available:../ssl/statem/statem_clnt.c:3737:No ciphers enabled for
max supported SSL/TLS version
```

**Comment by mattcaswell** [comment]
Like this:

````
% openssl s_client -cipher "NULL-SHA:@SECLEVEL=0" -no_tls1_3 -connect localhost:44330
````

The default security level does not support that ciphersuite, so you need to reduce it to security level 0.

Incidentally, if this is just for learning purposes, you can also get a lot of information about what is happening "on the wire" using the "-trace" option to s_client.

**Comment by ClemFandango2** [comment]
Sorry, something wrong. I just tried.
Server side:

```
% openssl s_server -key key.pem -cert cert.pem -accept 44330 -www -cipher NULL-SHA
Using default temp DH parameters
ACCEPT
4047D5F28D7F0000:error:0A0000C1:SSL routines:tls_post_process_client_hello:no shared cipher:../ssl/statem/statem_srvr.c:2220:
```

Client side:
```
% openssl s_client -cipher "NULL-SHA:@SECLEVEL=0" -no_tls1_3 -connect localhost:44330
CONNECTED(00000003)
40D7BE704F7F0000:error:0A000410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:../ssl/record/rec_layer_s3.c:1584:SSL alert number 40
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 118 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1695127173
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```
(Client exists without allowing me to enter the "GET /" string.)

**Comment by mattcaswell** [comment, marked as answer]
You also need the SECLEVEL bit on the server side if connecting to s_server:

````
% openssl s_server -cipher "NULL-SHA:@SECLEVEL=0" -www -key key.pem -cert cert.pem -accept 44330 
````

**Reply by ClemFandango2** [reply]
Marvelous! Thank you so much!

Quick question: is it possible to turn on NULL cipher in Apache mod_ssl without recompiling it?
Manual says:
```
aNULL, eNULL and EXP ciphers are always disabled

Beginning with version 2.4.7, null and export-grade ciphers are always disabled, as mod_ssl unconditionally adds !aNULL:!eNULL:!EXP to any cipher string at initialization.
```
https://httpd.apache.org/docs/2.4/mod/mod_ssl.html

**Reply by mattcaswell** [reply]
Sorry, I don't really know Apache config.

**Comment by ClemFandango2** [comment]
During that, packets are not encrypted, but RSA-encrypted master key is sent to server and client adds MAC (verify data) to packets sent.
I wonder, is it also possible to set auth algo to NULL? So that MAC wouldn't be required by server?
Is this what is called eNULL and aNULL?
Is it possible to use TLS_NULL_WITH_NULL_NULL for all packets?
TIA!
